### PR TITLE
pidstore: Optimize texkey generation

### PR DESCRIPTION
**Pull request by pazembrz**

### CODE REVIEW

File: backend/inspirehep/pidstore/providers/texkey.py [hunk 1]
```diff
@@ -75,41 +86,20 @@ def get_texkey_with_random_part(cls, texkey):
...
-            and texkey.object_uuid == uuid
-            and texkey.status == PIDStatus.REGISTERED
-        ):
-            return True
-        elif texkey:
-            raise PIDAlreadyExistsError(texkey.pid_type, texkey.pid_value)
-        return False
-
     @classmethod
     def create_or_update_texkey_in_pidstore(cls, texkey, uuid):
```
Comment (michamos) [inline backend/inspirehep/pidstore/providers/texkey.py:101]: I didn't catch this earlier, but this is a misnomer: it doesn't actually create a texkey in the pidstore.

File: backend/inspirehep/pidstore/providers/texkey.py [hunk 2]
```diff
@@ -124,42 +114,71 @@ def create(
...
+                    status=PIDStatus.REGISTERED,
+                    object_type=object_type,
+                    object_uuid=object_uuid,
+                    **kwargs,
+                )
+            return
+
+        first_part = cls.build_texkey_first_part(data)
+        if not first_part:
+            raise TexkeyCannotGenerateFirstPart
```
Comment (michamos) [inline backend/inspirehep/pidstore/providers/texkey.py:129]: doesn't `build_texkey_first_part` already raise? if you're afraid you could get empty content there and want to check here, better remove the exception from that method.

File: backend/inspirehep/pidstore/providers/texkey.py [hunk 3]
```diff
@@ -124,42 +114,71 @@ def create(
...
+        second_part = cls.build_texkey_second_part(data)
+        if not second_part:
+            raise TexkeyCannotGenerateSecondPart
```
Comment (michamos) [inline backend/inspirehep/pidstore/providers/texkey.py:132]: same

File: backend/inspirehep/pidstore/providers/texkey.py [hunk 4]
```diff
@@ -124,42 +114,71 @@ def create(
...
+
+        new_texkey = f"{first_part}:{second_part}"
+        texkeys_from_metadata = data.get("texkeys", [])
+
+        if len(texkeys_from_metadata) == 0 or not texkeys_from_metadata[0].startswith(
+            new_texkey
+        ):
```
Comment (michamos) [inline backend/inspirehep/pidstore/providers/texkey.py:139]: this is hard to follow. I think you should invert the `if` so you can return early, and don't need to indent so much for the main logic.

File: backend/inspirehep/pidstore/providers/texkey.py [hunk 5]
```diff
@@ -124,42 +114,71 @@ def create(
...
+                    with db.session.begin_nested():
+                        new_texkey_pid = super().create(
+                            pid_value=full_texkey_generated,
+                            status=PIDStatus.REGISTERED,
+                            object_type=object_type,
+                            object_uuid=object_uuid,
+                            **kwargs,
+                        )
+                except IntegrityError:
+                    LOGGER.error(
```
Comment (michamos) [inline backend/inspirehep/pidstore/providers/texkey.py:156]: warning?

File: backend/inspirehep/pidstore/providers/texkey.py [hunk 6]
```diff
@@ -38,20 +44,25 @@ def sanitize(cls, name):
...
-            return cls.sanitize(
-                ParsedName.loads(data["authors"][0].get("full_name")).last
+        full_name = get_value(data, "authors[0].full_name")
+        if full_name:
+            parsed_name = ParsedName.loads(full_name)
+            parsed_name = (
+                parsed_name.last if len(parsed_name) > 1 else full_name.split(",")[0]
             )
+        else:
+            parsed_name = None
```
[the remark this answers is not in the document]
Comment (pazembrz) [reply backend/inspirehep/pidstore/providers/texkey.py:54]: What this `first_author_name` is?

File: backend/inspirehep/pidstore/providers/texkey.py (hunk 1 above)
Comment (pazembrz) [reply backend/inspirehep/pidstore/providers/texkey.py:101]: :+1:

File: backend/inspirehep/pidstore/providers/texkey.py (hunk 2 above)
Comment (pazembrz) [reply backend/inspirehep/pidstore/providers/texkey.py:129]: :+1:

File: backend/inspirehep/pidstore/providers/texkey.py (hunk 3 above)
Comment (pazembrz) [reply backend/inspirehep/pidstore/providers/texkey.py:132]: :+1:

File: backend/inspirehep/pidstore/providers/texkey.py (hunk 4 above)
Comment (pazembrz) [reply backend/inspirehep/pidstore/providers/texkey.py:139]: :+1:

File: backend/inspirehep/pidstore/providers/texkey.py (hunk 5 above)
Comment (pazembrz) [reply backend/inspirehep/pidstore/providers/texkey.py:156]: :+1:

File: backend/inspirehep/pidstore/providers/texkey.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (michamos) [reply backend/inspirehep/pidstore/providers/texkey.py:54]: instead of `full_name`, `parsed_name`

[the remark this answers is not in the document]
Comment (pazembrz) [reply backend/inspirehep/pidstore/providers/texkey.py:54]: I just tried to do this and it looks for me like it's much less readable :/

**Review by michamos — APPROVED** [review]
Small typo, but looks ok :+1:
